Features/rental assistance part 2 #158503202

fddayan opened the pull request.

fddayan committed:
WIP

fddayan committed:
[#155196535] Show other field when other selected

fddayan committed:
[#155196535] delete pull right. Tests fixed

callaghanc committed:
[#155196535] Sets size on Add Rental Assistance button

callaghanc committed:
Merge branch 'master' into features/rental-assistance-part-1-#155196535

callaghanc committed:
Merge branch 'master' into features/rental-assistance-part-1-#155196535

callaghanc committed:
[#155196535] Makes a small spelling correction

fddayan committed:
WIP

fddayan committed:
Merge branch 'features/rental-assistance-part-1-#155196535' of github.com:Exygy/sf-dahlia-lap into features/rental-assistance-part-1-#155196535

fddayan committed:
[#155196535] Behavior improved

fddayan committed:
[#155196535] Snapshots updated

fddayan committed:
[#155196535] Using original in expandable table

fddayan committed:
[#155196535] Lint updated

callaghanc committed:
[#155196535] Does layout and spelling corrections

callaghanc committed:
[#155196535] Corrects some field mappings and displays

fddayan committed:
[#155196535] Fixed yes no radio button

fddayan committed:
[#155196535] Snapshots updated

callaghanc committed:
[#155196535] Adds Salesforce call mocks and updates some naming

fddayan committed:
[#158503202] Adding RentalAssistance model end to end

callaghanc committed:
[#158503202] Completes saving of rental assistances; refactors leases in Rails

callaghanc committed:
[#158503202] Adds lease management when dealing with rental assistances

callaghanc committed:
[#158503202] Adds existing rental assistances to supp app on page load

callaghanc committed:
[#158503202] Adds ability to delete a rental assistance

fddayan committed:
[#155196535] Adding validations

fddayan committed:
[#155196535] Adding primary applicant

fddayan committed:
[#155196535] Adding primary applicant

fddayan committed:
[#155196535] Snapshots updated

fddayan committed:
[#155196535] Validate only type of assistance

callaghanc committed:
[#158503202] WIP - beginning to write e2e test for rental assistance

fddayan committed:
[#155196535] Adding tests

fddayan committed:
[#158503202] Merged with rental assistance part 1

fddayan committed:
SAVEPOINT

callaghanc committed:
[#155196535] Does a bit of cleanup on snapshot tests

callaghanc committed:
Merge branch 'master' into features/rental-assistance-part-1-#155196535

callaghanc committed:
[#155196535] Does a bit of linting

callaghanc commented on config/routes.rb:
```diff
@@ -27,24 +29,28 @@
+
       resources :preferences, only: %w[update]
+
+      resources :rental_assistances, path: '/rental-assistances', only: %w[create update destroy]
```
This is the only important changed line relevant to rental assistances in this file. The rest of the changes are just tidying and alphabetization.

callaghanc committed:
[#155196535] Improves new rental assistance form panel toggle var name

callaghanc committed:
[#158503202] Adds rspec tests for lease and rental assistance models

callaghanc committed:
Merge branch 'features/rental-assistance-part-1-#155196535' into features/rental-assistance-part-2-#158503202

callaghanc commented on app/javascript/components/supplemental_application/sections/RentalAssistance.js:
```diff
@@ -137,23 +138,24 @@ const validateError = (values) => {
     <Form onSubmit={onSave} defaultValues={values} validateError={validateError}>
       {formApi => (
-        <div className='app-editable expand-wide scrollable-table-nested'>
+        <div className='app-editable expand-wide scrollable-table-nested rental-assistance-form'>
```
This class and the others added below this in this file are added to enable us to test. The e2e tests need identifiers to be able to locate and interact with the rental assistance form components.

callaghanc committed:
[#158503202] Makes e2e test browsers close at end of test

callaghanc commented on spec/javascript/e2e/ApplicationNewPage.e2e.js:
```diff
@@ -34,7 +34,7 @@ describe('ApplicationNewPage', () => {
     expect(values).toContain(DATE_OF_BIRTH)
 
-    browser.close()
+    await browser.close()
```
All of these `await`s that you see added in this PR before the `browser.close()`s in our e2e tests are to ensure that we actually wait for the browser to close! Not having these awaits is what was causing the browsers in the e2e tests to never actually close and causing the whole test process to hang for a long time after finishing the tests.

fddayan committed:
[#158503202] Making e2e tests passed with new strategy

fddayan committed:
[#158503202] Merged with latest

fddayan committed:
[#155196535] Do not render other assistance name  if other option is not selected

callaghanc committed:
[#155196535] Updates rental assistance snapshot tests

fddayan committed:
[#155196535] Reset other assistance name if type of assistance changed

callaghanc committed:
Merge branch 'master' into features/rental-assistance-part-1-#155196535

fddayan committed:
[#155196535] Comment removed

fddayan committed:
Merge branch 'features/rental-assistance-part-1-#155196535' of github.com:Exygy/sf-dahlia-lap into features/rental-assistance-part-1-#155196535

fddayan committed:
[#158503202] Merged with part 1

fddayan committed:
[#158503202] 'error' class added back.

callaghanc commented:
This PR is code-complete, but we are going to hold off on reviewing this until part 1 is fully QA'ed and merged to master, so this PR will remain WIP for now.

fddayan committed:
[#158503202] Setting clasaName in field

fddayan committed:
SAVEPOINT

callaghanc committed:
Merge branch 'master' into features/rental-assistance-part-2-#158503202

callaghanc committed:
[#158503202] Standardizes supp app page e2e test use of headless option

callaghanc committed:
[#158503202] Updates rental assistance tests

callaghanc committed:
[#158503202] Updates rental assistance snapshot

callaghanc committed:
[#158503202] WIP - further develops the rental assistance update e2e test

fddayan committed:
[#158503202] Show buttons as loading

fddayan committed:
[#158503202] Merged with orgigin

callaghanc committed:
[#158503202] Add e2e test for deleting rental assistance

callaghanc committed:
[#158503202] Removes extra wrapper div on new rental assistance form

callaghanc committed:
[#158503202] Adds missing arrow-down.png file

callaghanc committed:
[#158503202] Stops rental assistance panel from closing on failed update or delete

callaghanc commented:
This PR is ready for review.

akegan commented on app/javascript/components/supplemental_application/SupplementalApplicationPage.js:
```diff
@@ -182,62 +183,85 @@ class SupplementalApplicationPage extends React.Component {
+      })
+      return true
+    } else {
+      Alerts.error()
```
does this already display a message or do you have to pass one in?

callaghanc replied:
this triggers our standard window alert, which has the message "Oops! Looks like something went wrong. Please try again."

akegan commented on app/javascript/components/supplemental_application/SupplementalApplicationPage.js:
```diff
@@ -182,62 +183,85 @@ class SupplementalApplicationPage extends React.Component {
+      Alerts.error()
+      this.setState({ rentalAssistanceLoading: false })
+      return false
+    }
```
do you think that it'd be possible to condense the create/update/destroy or are there enough slight differences that it's not worth it?

callaghanc replied:
there are enough differences in the post-request steps that it's probably not worth squishing them into a single method; however, perhaps the loading state management and error alert could be DRYed a bit - @fddayan what are your thoughts?

fddayan replied:
I would keep a method for each action. I agree we would DRY up those methods.

callaghanc replied:
@akegan Fed has added a commit showing a way to draw out the common parts of the rental assistance action functions - please review and let us know if you'd like to use that new way or keep it as it was before.

akegan replied:
I'm ambivalent! Either way works. Up to you guys!

callaghanc replied:
Ok let's leave it as it is now

akegan commented on app/javascript/components/supplemental_application/SupplementalApplicationPage.js:
```diff
@@ -332,14 +358,15 @@ const setApplicationsDefaults = (application) => {
     units: mapList(mapUnit, units),
-    availableUnits: mapList(mapUnit, availableUnits)
+    availableUnits: mapList(mapUnit, availableUnits),
+    rentalAssistances: rentalAssistances
```
why doesn't rentalAssistances need to use mapList?

callaghanc replied:
because the rental assistances already come in in domain format! 🎉 thanks to our refactor work, the SOQL rental assistance service uses SOQL to fetch the rental assistances, and then translates them to domain before passing them up to the controller. that way the controller can just hand the frontend domain-formatted rental assistances. the ultimate goal here is to do away with all use of `mapList` in the frontend. and then, even further then that, once that's done and the frontend is entirely in domain, we can consider whether we'd like to convert domain format from snake case to camel case - but that's a little further off yet.

akegan commented on app/services/force/soql/rental_assistance_service.rb:
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
```
so many new clean and organized services! 🎉

callaghanc replied:
😍

akegan commented on spec/fixtures/model/force/lease/lease_domain.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "application_id": "a0o0P00000FFc2yQAD",
+  "id": "a130U0000001n4wQAA",
+  "leaseStartDate": "2028-01-01",
```
I thought domain was snake_case? I believe other domain fixtures are.

callaghanc replied:
yeah, for some reason all the already-existing lease domain fields were in camel case when i came to this story. i think you and @chrisdolendo did the previous lease work - does anyone recall why this is? i'm fine with changing them to snake case in this PR if there's no problem with doing that. even though we may ultimately decide that we want to convert domain to camel case down the line, until then it would be better to maintain consistency

akegan replied:
yeah probably my fault . If you're up for it I think we should make it consistent. If you'd prefer, I can followup in a new PR after this one

callaghanc replied:
i'll do it in this PR, i got most of it done already during retro and sprint planning 😀

callaghanc replied:
done!

akegan commented on spec/fixtures/model/force/lease/lease_salesforce_from_domain.json:
```diff
@@ -0,0 +1,12 @@
+{
```
does this file name need the "from_domain" or is it just lease_salesforce?

callaghanc replied:
we specify both the origin format and the result format, because the content of a translated object will depend on where that object originated from. objects can have fields that are only present when they are fetched from salesforce, or fields that are only present when they are fetched from the custom api. if i fetch, say, an application from the custom api, and then translate it to domain, it will only have the domain correlates of the custom api fields. if i fetch that same application via SOQL and translate it to domain, it will have the domain field correlates for those salesforce fields, which are a slightly different set of fields from what is returned by the custom api. this is a lot more relevant for application and preference objects, but i wanted to maintain consistency when testing all objects.

akegan commented on spec/fixtures/model/force/rental_assistance/rental_assistance_salesforce_from_domain.json:
```diff
@@ -0,0 +1,8 @@
+{
```
same here re: filename

akegan commented on spec/javascript/e2e/SupplementalApplicationPage.e2e.js:
```diff
@@ -39,7 +39,140 @@ describe('SupplementalApplicationPage', () => {
+      '.rental-assistance-new-form .rental-assistance-type option:nth-child(2)',
+      e => [e.value, e.textContent]
+    )
+    const recurring = 'Yes'
```
nit: this section might be bit clearer if the page.$evals are grouped before or after the recurring/amount.

callaghanc replied:
i wanted to keep them in the order that they're filled in on the page - especially for debugging, it makes it a lot easier to see "ok, we wanted to use this value for this field, then this value for the next....and the test got to this field, then the next..."

akegan commented on spec/javascript/e2e/SupplementalApplicationPage.e2e.js:
```diff
@@ -39,7 +39,140 @@ describe('SupplementalApplicationPage', () => {
+    expect(newTableSize).toEqual(prevTableSize + 1)
+
+    // Check that the last rental assistance in table has the values we saved
+    const lastRentalAssistanceSelector = '.rental-assistances > tbody > .tr-expand:nth-last-child(2)'
```
why is it nth-last-child(2) if it's the last row?

callaghanc replied:
because it's not actually the last table row, it's only the last visible row - there's a hidden row right after it that is its corresponding expandable panel

akegan commented on spec/javascript/e2e/SupplementalApplicationPage.e2e.js:
```diff
@@ -21,7 +21,7 @@ describe('SupplementalApplicationPage', () => {
 
```
at some point soon we may want to think about breaking this up into multiple files so it doesn't get too long. For now, do you think it might help to add another layer of "describe" to help organize the e2e tests a bit?

callaghanc replied:
yeah i've also been thinking that - i wonder if there's any problem with doing that breaking up now? @fddayan @chrisdolendo how would you guys feel about breaking up the supp app e2e file now?

fddayan replied:
👍I'm ok with splitting the file if it's growing. We could re think the naming convention we use for e2e tests. Instead of using the Page as the main test we can can `SupplementalApplicationPage/feature_name.e2e.test` 
like 
`SupplementalApplicationPage/manage_rental_asistances.e2e.test` .... or any name

callaghanc replied:
@akegan Fed and I have broken apart the supp app e2e tests in a couple of recent commits - please review.

akegan reviewed (commented):
I think these are all only comments/questions. Looks great! I really appreciate all the cleanup on the backend and the e2e tests!

also hooray for adding the arrow down! 🎉

akegan reviewed (approved):
looks good. 

Only outstanding things are possibly
1. Updating lease to have snake case domain model
1. Splitting up or organizing supplemental app tests

But it all looks good to me!

callaghanc committed:
[#158503202] Standardizes lease domain field names

callaghanc commented:
I've updated this PR with standardized lease domain field names. This PR is awaiting some responses from @fddayan and/or @chrisdolendo on some questions.

fddayan committed:
[#158503202] Dry out rental assistance actions. Spli Supp page e2e into features

callaghanc committed:
[#158503202] Completes translation of lease to domain in Rails

Removes need to map lease in the React app. Also cleans up the e2e test organization a bit.

callaghanc committed:
[#158503202] Adds missing preference app member ID mapping

The preference model did not have a complete mapping for the application member ID. This commit adds that. It also updates related rspec tests with app member IDs, and with some numbers that changed to floats in Salesforce.

callaghanc commented:
@akegan Fed and I have added a few commits to this PR. In addition to addressing the two items you listed as outstanding in your last comment, these commits also do a bit of lease mapping work and correct a field mapping issue with preferences. So please review the commits since you last reviewed and let us know your thoughts.

callaghanc committed:
[#158503202] Uses new custom API endpoint for deleting rental assistances

Using the standard REST API, it is not possible for a user with lower permissions to delete a rental assistance that was created by a user with higher permissions. The custom API endpoint provided by Vertiba allows a user with lower permissions to delete a rental assistance even if it was added by a user with higher permissions.

akegan commented on spec/javascript/e2e/SupplementalApplicationPage/preferences.e2e.js:
```diff
@@ -0,0 +1,61 @@
+import puppeteer from 'puppeteer'
```
small nit: filenames aren't super consistent for these new broken-out things.

action_buttons
preferences
manage_rental_assistance
status_history

I think it might be a bit clearer if we pick a convention and stick with it. 

I'd prefer if we keep them to domain-related topics, e.g. 
-preferences
-rental_assistance
-status_update (combine status history and action_buttons)

callaghanc replied:
sounds good to me, done!

callaghanc committed:
[#158503202] Corrects response from rental assistance action handler

Also corrects rental assistance type names back to their original Salesforce picklist capitalization, and adds check for presence of lease in rental assistances controller.

callaghanc commented on app/javascript/components/supplemental_application/sections/RentalAssistance.js:
```diff
@@ -26,9 +27,9 @@ const typeOfAssistance = [
+  'Section 8 HCV (tenant) voucher',
   'Self-Help for the Elderly',
-  'VASH Voucher',
+  'VASH voucher',
```
@jrubenoff I've changed the capitalization on these two options back to their original Salesforce picklist capitalizations, so that they will exactly match the picklist values.

jrubenoff replied:
That's unfortunate, but OK!

callaghanc committed:
[#158503202] Renames supp app e2e tests

akegan committed:
Merge branch 'master' into features/rental-assistance-part-2-#158503202

akegan commented:
@callaghanc I've merged latest master into this 🤞the tests will pass and we can merge this in
